docs/GroupDefs.hpp

```diff
@@ -370,145 +370,263 @@ that load balancing calls may result in Array elements being moved.
+to return a DataBox with the same type that is passed in and also terminated
+the algorithm:
+
+\snippet Test_AlgorithmParallel.cpp return_with_termination
+
+Notice that we again return a reference to the DataBox, which is done to avoid
+any copying. After an algorithm has been terminated it can be restarted by
+passing `false` to the `set_terminate` method followed by calling the
+`perform_algorithm` or `receive_data` methods.
+
+The third optional element in the returned `std::tuple` is a `size_t` whose
+integer value corresponds to the index of the action to be called next in the
+`action_list`. The metafunction `tmpl::index_of<list, element>` can be used to
+get an `integral_constant` with the value of the index of the element `element`
+in the typelist `list`. For example,
+
+\snippet Test_AlgorithmCore.cpp out_of_order_action
+
+Again a reference to the DataBox is returned, while the termination `bool` and
+next action `size_t` are returned by value. The metafunction call
+`tmpl::index_of<ActionList, iterate_increment_int0>::%value` returns a `size_t`
+whose value is that of the action `iterate_increment_int0` in the `action_list`.
+The indexing of actions in the `action_list` starts at `0`.
+
+Iterable actions are invoked as part of the algorithm and so the only way
+to request they be invoked is by having the algorithm run on the parallel
+component. The algorithm can be explicitly triggered by call the
+`perform_algorithm` method:
+
+\snippet Test_AlgorithmCore.cpp perform_algorithm
+
+The algorithm is also evaluated by calling the `receive_data` function, either
+on an entire array or singleton (this does a broadcast), or an on individual
+element of the array. Here is an example of a broadcast call:
+
+\snippet Test_AlgorithmParallel.cpp broadcast_to_group
+
+and of calling individual elements:
+
+\snippet Test_AlgorithmParallel.cpp call_on_indexed_array
+
+The `receive_data` function always takes a `ReceiveTag`, which is set in the
+actions `inbox_tags` type alias as described above.  The first argument is the
+temporal identifier, and the second is the data to be sent.
 
 #### 3. Reduction %Actions
 
-Blah
+Finally, there are reduction actions which are used when reducing data over an
+array. For example, you may want to know the sum of a `int` from every
+element in the array. You can do this by using Charm++'s built-in reduction
+function for summing `int`s as follows:
+
+\snippet Test_AlgorithmReduction.cpp contribute_to_reduction_example
+
+in the array. This reduces over the parallel component
+`ArrayParallelComponent<Metavariables>`, reduces to the parallel component
+`SingletonParallelComponent<Metavariables>`, and calls the action
+`singleton_reduce_sum_int`. The `CkReduction::sum_int` argument tells Charm++
+that you want to sum integers instead of some other type or operation. The
+reduction action is:
+
+\snippet Test_AlgorithmReduction.cpp singleton_reduce_sum_int
```

Comment:
make `value` const

Reply:
The prefix `singleton` doesn't make it clear that this is an action. Just remove the prefix, or put the keyword `action` somewhere in the function name. Also the name suggests that it does the reduction, but it is only the receiver. And it is a type, so it should be camel case. So perhaps better call it `ProcessReducedSumOfInts` or similar.